### ISSUE TITLE
Proposal: rearrange notebooks to focus on "get started", "solve a problem" and "reproduce a paper" separately

### ISSUE THREAD

**Issue by huonw** [stellargraph/stellargraph#910]
#### Current state

Currently we have an awesome set of extensive notebooks, for each algorithm. [At the top level](https://github.com/stellargraph/stellargraph/tree/8e3c9b64983d57afcdd7780f31de75436c7df2de/demos), these are grouped by topic like ["node classification"](https://github.com/stellargraph/stellargraph/tree/8e3c9b64983d57afcdd7780f31de75436c7df2de/demos/node-classification). Then, they are grouped by algorithm name/abbreviation. This level of grouping typically has a README that gives a brief summary of what each of the subfolders do as an unstructured list.

The current set of notebooks is serving a few purposes:

1. teaching how to use stellargraph (e.g. load/preprocess data)
2. describing each algorithm and how it works to solve some task (e.g. the exact way to set up the models and train)
3. showing how to work with the results of the given task (e.g. doing predictions, visualising embeddings with PCA/TSNE)
4. validating the algorithms are functioning as described, by reproducing the results of the paper

1 is done in every notebook, for instance, most of the notebooks working with Cora have essentially the same content in first 4 sections (Loading the CORA network, Splitting the data, Converting to numeric arrays, Create the StellarGraph object), e.g.:

- https://github.com/stellargraph/stellargraph/blob/8e3c9b64983d57afcdd7780f31de75436c7df2de/demos/node-classification/gcn/gcn-cora-node-classification-example.ipynb
- https://github.com/stellargraph/stellargraph/blob/8e3c9b64983d57afcdd7780f31de75436c7df2de/demos/node-classification/sgc/sgc-node-classification-example.ipynb

3 is also done in every notebook, even though there's a very large amount of overlap: node attribute prediction generally give embedding vectors as a layer, and these can be used to do PCA/TSNE visualisation. The notebooks above share "Making predictions with the model" and "Node representations"/"Node embeddings" sections.

4 seems to be done incidentally as part of the teaching notebooks, e.g. [`rgcn-aifb-node-classification-example.ipynb `](https://github.com/stellargraph/stellargraph/blob/8e3c9b64983d57afcdd7780f31de75436c7df2de/demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb) works with the same as the paper, and (implicitly) matches the test set validation to [Table 2 in the paper](https://arxiv.org/pdf/1703.06103.pdf).

#### Discussion

This current state seems great for user looking at a notebook for a specific algorithm on a specific task in isolation, without comparing/contrasting any other algorithms (or tasks). 

Unfortunately, there seems to be quite a few downsides to this:

- it seems to work best when a user knows which algorithm is interesting for them, which seems unlikely for many data scientists with graph data. For instance, someone with a task like "node classification on a heterogenous graph", can get to the `node-classification` folder, but then has to read the README carefully to work out that they can try looking at `/hinsage`.
- there's a lot of duplication with 1 and 3, which likely gets in the way in two cases:
  - maintaining the library and making updates requires doing changes across many notebooks
  - users familiar with the basic infrastructure have to wade through the duplication (e.g. someone doing a comparison of algorithms)
- doing 4 implicitly means we don't have a "definite" or exact reproduction of papers. If we did have this, that tried to get very close to many/all of the results in each paper, downstream users (such as people writing papers) could use these as easily-reproduced reference results
- it doesn't follow guidelines like https://www.divio.com/blog/documentation/, since each notebook is a mix of a Tutorial (1 in the list above), Explanation/Reference (2) and How-To guide (3), and even "integration test" (4)

#### Possible improvement

We could retain the broadly the same content, but split differently: instead of having them split where each notebook covers many aspects of documentation for a specific algorithm, we could try to have:

1. basic Tutorials: `getting_started.ipynb` that shows how to load a dataset like Cora and create a stellargraph, and maybe `machine_learning.ipynb` that works with the same dataset (but without all the manually parsing code) and builds a basic model to train and predict with
2. How-To guides for specific tasks: "node classification on a homogeneous graph with attributes", "node classification on a homogeneous graph without attributes", "link prediction on a homogeneous graph ...", "link prediction on a knowledge graph ...". These could run multiple algorithms. A data scientist with a specific problem would then go straight to the notebook for that problem and be able to see all the options. For example, for the first, there might be a notebook that does:

  - Prepare data (load, do splits, ...) with a very brief explanation, mostly in the form of references to other notebooks or documentation
  - Train GCN, e.g. `gcn_gen = ...; gcn_model = ...; gcn_model.fit(gcn_gen.flow(...))`
  - Train GAT, e.g. `gat_gen = ...; gat_model = ...; gat_model.fit(gat_gen.flow(...))`
  - Train GraphSAGE, e.g. `graphsage_gen = ...; graphsage_model = ...; graphsage_model.fit(graphsage_gen.flow(...))`
  - ...
  - Summarise performance of all models
  - Demonstrate predictions with one model (since they all have the same output shape), e.g. `graphsage_model.predict(graphsage_model.flow(...))`
  - Potentially, demonstrate looking at node embeddings for one model
3. paper reproductions: have a single notebook for each algorithm that reproduces the paper (this might be arranged like `demos/details/gcn.ipynb`, `demos/details/watch-your-step.ipynb`, ..., or some folder name other than `details`).
4. Reference and/or Explanation material, in the paper reproductions (3), or the in the read-the-docs documentation, or where it talks in more detail about how the algorithm works and how it works in practice (e.g. what changes had to be made to the paper's description to get it working).

(I'm not sure about 4.)

### Done Checklist

- [ ] Produced code for required functionality
- [ ] Tests written and coverage checked
- [ ] Code review performed
- [ ] Documentation on Google Docs  (if applicable)
- [ ] Documentation in repo
- [ ] Version number reflects new status
- [ ] CHANGELOG.md updated
- [ ] Team demo

**Comment by kjun9** [comment]
Nice!

We could probably get started with doing improvements 1 and 2, but we may need to relax 3 since we don't currently have a notebook for each algorithm that reproduces the results from the paper and this might not be straightforward:

* some papers are vague enough in describing their experimental setup that it could be difficult to try and reproduce their results
* @habiba-h mentioned to me that some of our implementation of algorithms currently in the library may not exactly follow the reference paper (our interpretation of vagueness in the paper, or maybe some things seemed like it could be improved). 

Maybe improvements 3 and 4 could be combined, so that depending on the situation, we don't necessarily have to reproduce the reference experiment, and instead can just think of a way to demonstrate what the algorithm is useful for and what kind of patterns it is capable of discovering. Thoughts?

And we also have the issue @PantelisElinas raised:
> Having long notebooks that demonstrate several algorithms in one (see for example the Twitter use-case notebook) can be confusing for someone new to this specialisation as there is too much code in to follow. There are also several places one needs to check for the algorithms selected (if statements) in order to make things work correctly; this could make the multi-algorithm demo notebooks even more difficult to parse for a beginner.
 
IMO this refactoring can at least get us one notebook with slightly more code rather than lots of notebooks that have lots of duplication, which I still think would be a big improvement. But I guess we should keep in mind to make sure the how-to guides for specific tasks are as simple as possible.

**Comment by huonw** [comment]
> we may need to relax 3 since we don't currently have a notebook for each algorithm

Yeah, I think all of this would be aspirational rather than a hard requirement.

> some papers are vague enough in describing their experimental setup that it could be difficult to try and reproduce their results

Yeah, definitely, but we can at least have something.

> @habiba-h mentioned to me that some of our implementation of algorithms currently in the library may not exactly follow the reference paper (our interpretation of vagueness in the paper, or maybe some things seemed like it could be improved).

I think it's still worth reproducing the experiments done in the paper, as much as possible. For vagueness, this will give us some assurance that we're interpreting it approximately correctly if our numbers are close, and, for "could be improved", it'll help tell us whether they were actual improvements (or, at least, not de-provements).

It'll also give us a place to discuss these differences (although these are potentially best discussed in the read-the-docs documentation).

> But I guess we should keep in mind to make sure the how-to guides for specific tasks are as simple as possible.

Yeah, I agree. I had that concern in mind with the suggestion to have the notebook run all of the algorithms in series, unconditionally, so that there's no `if` statements to manage generators (etc.): each algorithm does all its own set-up in its own section. That said, without trying it out, I'm not sure if this will alleviate that issue enough.

**Comment by timpitman** [comment]
This rework of demos would definitely make StellarGraph more accessible for new users, and help clean up our sprawling demos.

Also worth considering is whether we should merge content from our repo at "stellar-practical-ml-on-graphs" which does do part of 1.

**Comment by timpitman** [comment]
I've split task 1. (teaching StelarGraph) into #992
